src/passes/EncloseWorld.cpp

```diff
@@ -0,0 +1,154 @@
+// useful for fuzzing (in that a random module is usually very incomptable with
+// closed world, with most types being public and hence unoptimizable, but
+// running this pass makes as many as we can fully private).
+//
+// The fixup we do is to find references sent out/received in, and to
+// externalize / internalize them. For example, this export:
+//
+//  (func $refs ("export "refs") (param $x (ref $X)) (result (ref $Y))
+//
+// would have the following function exported in its place:
+//
+//  (func $refs-closed ("export "refs") (param $x externref) (result externref)
+//    (extern.convert_any
+//      (call $refs
+//        (ref.cast (ref $X)
+//          (any.convert_extern
+//            (local.get $x))))))
+//
+
+#include "ir/names.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+struct EncloseWorld : public Pass {
+  void run(Module* module) override {
+    // Handle exports.
+    // TODO: Non-function exports.
+    std::vector<std::unique_ptr<Export>> newExports;
+    for (auto& ex : module->exports) {
+      if (ex->kind == ExternalKind::Function) {
+        auto* func = module->getFunction(ex->value);
+        // If this opens up types, replace it with an enclosed stub.
+        if (opensTypes(func)) {
+          auto stubName = makeStubStubForExport(func, module);
+          ex->value = stubName;
+        }
+      }
+    }
+    for (auto& ex : newExports) {
+      module->addExport(std::move(ex));
+    }
+
+    // TODO: Handle imports.
+  }
+
+private:
+  // Whether a type is an "open" ref, that is, a type that closed-world would
+  // consider to keep things public and prevent some amount of closed-world
+  // optimizations.
+  bool isOpenRef(Type t) {
+    // Only externref keeps things closed, and we must ignore things that
+    // cannot be converted to/from it (like funcrefs), so we can just check for
+    // the top type being any.
+    return t.isRef() && t.getHeapType().getTop() == HeapType::any;
+  }
+
+  // Whether a function causes types to be open.
+  bool opensTypes(Function* func) {
+    for (const auto& param : func->getParams()) {
+      if (isOpenRef(param)) {
+        return true;
+      }
+    }
+    // TODO: Handle tuple results.
+    return isOpenRef(func->getResults());
+  }
+
+  // Make an enclosed stub function for an exported function, and return its
+  // name.
+  Name makeStubStubForExport(Function* func, Module* module) {
+    // Pick a valid name for the stub we are about to create.
+    auto stubName = Names::getValidFunctionName(
+      *module, std::string("stub$") + func->name.toString());
+
+    // Create the stub.
+    Builder builder(*module);
+    auto* stub = new Function();
+    stub->name = stubName;
+    stub->hasExplicitName = true;
+
+    // The stub's body is just a call to the original function, but with some
+    // conversions to/from externref.
+    auto* call = module->allocator.alloc<Call>();
```

Comment:
It would be nice to use `Builder` (or even `IRBuilder` 👀) rather than manually allocating expressions.

Reply:
Oh, sorry, I seem to have copied this old code and not noticed how old it was... Fixed.